blockchain/blockchain_test.go

```diff
@@ -1438,3 +1441,122 @@ func TestCheckBlockChainVersion(t *testing.T) {
+		types.TxValueKeyTo:       contractAddr,
+		types.TxValueKeyAmount:   big.NewInt(0),
+		types.TxValueKeyGasLimit: gasLimit,
+		types.TxValueKeyGasPrice: gasPrice,
+		types.TxValueKeyData:     data,
+	}
+	tx, err = types.NewTransactionWithMap(types.TxTypeSmartContractExecution, values)
+	assert.Equal(t, nil, err)
+
+	err = tx.SignWithKeys(signer, []*ecdsa.PrivateKey{key})
+	assert.NoError(t, err)
+
+	block.AddTx(tx)
+}
+
+// TestCallTraceChainEventSubscription tests if the method insertChain posts a chain event correctly.
+// Scenario:
+// 1. Deploy a contract
+//   sendKlay(n uint32, receiver address): send 1 peb to `receiver` address `n` times.
+// 2. Send a smart contract execution transaction
+func TestCallTraceChainEventSubscription(t *testing.T) {
+	// configure and generate a sample block chain
+	var (
+		gendb       = database.NewMemoryDBManager()
+		key, _      = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+		address     = crypto.PubkeyToAddress(key.PublicKey)
+		funds       = big.NewInt(100000000000000000)
+		testGenesis = &Genesis{
+			Config: params.TestChainConfig,
+			Alloc:  GenesisAlloc{address: {Balance: funds}},
+		}
+		genesis = testGenesis.MustCommit(gendb)
+		signer  = types.NewEIP155Signer(testGenesis.Config.ChainID)
+	)
+	db := database.NewMemoryDBManager()
+	testGenesis.MustCommit(db)
+
+	// create new blockchain with enabled internal tx tracing option
+	blockchain, _ := NewBlockChain(db, nil, testGenesis.Config, gxhash.NewFaker(), vm.Config{Debug: true, EnableInternalTxTracing: true})
+	defer blockchain.Stop()
+
+	// subscribe a new chain event channel
+	chainEventCh := make(chan ChainEvent, 1)
+	subscription := blockchain.SubscribeChainEvent(chainEventCh)
+	defer subscription.Unsubscribe()
+
+	// generate blocks
+	blocks, _ := GenerateChain(testGenesis.Config, genesis, gxhash.NewFaker(), gendb, 1, func(i int, block *BlockGen) {
+		// Deploy a contract which can trigger internal transactions
+		genInternalTxTransaction(t, block, address, signer, key)
+	})
+
+	// insert the generated blocks into the test chain
+	if n, err := blockchain.InsertChain(blocks); err != nil {
+		t.Fatalf("failed to process block %d: %v", n, err)
+	}
+
+	// compare the published chain event with the expected test data
+	select {
+	case ev := <-chainEventCh:
+		// a contract deploy tx and a contract execution tx
+		assert.Equal(t, 2, len(ev.InternalTxTraces))
+
+		// compare contract deploy result
+		assert.Equal(t, address, ev.InternalTxTraces[0].From)
+		assert.Equal(t, 0, len(ev.InternalTxTraces[0].Calls))
+		assert.Equal(t, "0x"+testCode, ev.InternalTxTraces[0].Input)
+		assert.Equal(t, fmt.Sprintf("0x%x", 100000000), ev.InternalTxTraces[0].Value)
+
+		// compare contract execution result
+		assert.Equal(t, address, ev.InternalTxTraces[1].From)
+		assert.Equal(t, 3, len(ev.InternalTxTraces[1].Calls))
+		assert.Equal(t, fmt.Sprintf("0x%x", 0), ev.InternalTxTraces[1].Value)
+	}
```

Comment:
You better add a timeout for this logic. If a chainEvent is not posted with some faults, the test will not be terminated.

Reply:
Applied.